larq_compute_engine/tflite/kernels/bconv2d.cc

```diff
@@ -199,6 +170,35 @@ TfLiteStatus Prepare(KernelType kernel_type,
   TF_LITE_ENSURE_EQ(context, NumDimensions(post_activation_multiplier), 1);
   TF_LITE_ENSURE_EQ(context, NumDimensions(post_activation_bias), 1);
 
+  conv_params->read_bitpacked_input = input->type == kTfLiteInt32;
+  conv_params->write_bitpacked_output = output->type == kTfLiteInt32;
+
+  if (conv_params->padding_type == TfLitePadding::kTfLitePaddingSame &&
+      conv_params->pad_value != 1 && conv_params->write_bitpacked_output) {
+    context->ReportError(context,
+                         "Writing bitpacked output is only supported with "
+                         "valid or one-padding.");
+    return kTfLiteError;
+  }
+
+  // Read the input dimensions. TF Lite has the same input format as TensorFlow:
+  // (B, H, W, Ci).
+  conv_params->batch = input->dims->data[0];
+  conv_params->input_height = input->dims->data[1];
+  conv_params->input_width = input->dims->data[2];
+  if (!conv_params->read_bitpacked_input) {
+    TF_LITE_ENSURE_EQ(context, conv_params->channels_in, input->dims->data[3]);
+  } else if (conv_params->channels_in == 0) {
```

Comment:
Since `0` is not the default argument in the converter anymore, I think we should check against null, or remove this check.

Reply:
channels in can never be `0` can it?

Reply:
0 is the default argument in the `conv_params` constructor, so the value will be 0 if it's not otherwise set (in `Init` we set the value, but only if it's non-null in the flatbuffer). The only reason this check would be relevent would be if you tried to run with an old model that was converted with an old converter and so didn't have the `channels_in` attribute set, but that's still probably something we should catch.

Reply:
On a related note, I don't think we have a proper way of versioning the ops/the converter?

Reply:
> On a related note, I don't think we have a proper way of versioning the ops/the converter?

Currently not, though builtin TFLite ops do. I think we should investigate if we can get some sort of versioning at some point.

Reply:
When registering an op you can provide a version, so from the runtime side thats very easy to do. I'm not sure how easy it is from the mlir/flatbuffer side.

Reply:
> I'm not sure how easy it is from the mlir/flatbuffer side.

I can take a look at it later this week. TF 2.2 added support for this in https://github.com/tensorflow/tensorflow/commit/42a4ae6cbef9505fd7f7f7f7d17e869da75b9583